### PR TITLE
Lab#4 Rodionov Nikita 381906-1

### PR DESCRIPTION
**Pull request by TexHik620953**

### CODE REVIEW

**Comment by TexHik620953** [comment]
@ggladilov , please review

File: llvm/include/llvm/Transforms/PeepholeOptimizationCourse/RemoveAddWithZeroNewPassManager.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef LLVM_TRANSFORMS_PEEPHOLE_REMOVE_ADD_WITH_ZERO_H
```
Comment (ggladilov) [inline llvm/include/llvm/Transforms/PeepholeOptimizationCourse/RemoveAddWithZeroNewPassManager.h:1]: Remove my commits from your PR

File: llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8Bytes/Obfuscate8BytesAdd.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                                    )
+                                )
+                            )
+                        )
+                    )
+                );
+
+                ReplaceInstWithInst(basicBlock->getInstList(), instruction, obfuscated);
+
+                LLVM_DEBUG(dbgs() << *binaryOperator << " -> " << *obfuscated << '\n');
```
Comment (ggladilov) [inline llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8Bytes/Obfuscate8BytesAdd.cpp:62]: What is it and what it does?

File: llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+
+#include "llvm/Support/raw_ostream.h"
+
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+using namespace llvm;
+
+bool isBinaryAdd(Instruction const* instruction) {
```
Comment (ggladilov) [inline llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp]: Make all free functions here to be of internal linkage

File: llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    return instruction->getOpcode() == Instruction::Add && instruction->getNumOperands() == 2;
+}
+
+bool isI8(Value const* operand) {
+    return (operand->getType()->isIntegerTy()) && (operand->getType()->getIntegerBitWidth() == 8);
+}
+
+bool withI8(Instruction const* instruction) {
```
Comment (ggladilov) [inline llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp]: It is easier to check bitwidth of operation itself:

```
binOp->getType()->isIntegerTy() && binOp->getType()->getIntegerBitWidth() == 8
```

File: llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	auto const& operands = instruction->operands();
+    return std::any_of(operands.begin(), operands.end(),
+                       [](auto const& operand) { return isI8(operand); });
+}
+
+bool isAddWithI8(Instruction const* instruction) {
+	return isBinaryAdd(instruction) && withI8(instruction);
+}
+
+void doObfuscateI8Add(Instruction* instruction) {
```
Comment (ggladilov) [inline llvm/lib/Transforms/PeepholeOptimizationCourse/Obfuscate8BytesNewPassManager/Obfuscate8BytesNewPassManager.cpp]: Could you please reuse that code between passes?

**Pull request closed without merging**